File: platforms/ios/Dezel/Dezel/Modules/Global/JavaScriptWebSocket.swift
/**
 * @class JavaScriptWebSocket
 * @super JavaScriptClass
 * @since 0.7.0
 */
public class JavaScriptWebSocket: JavaScriptClass, WebSocketDelegate {

	//--------------------------------------------------------------------------
	// MARK:Properties
	//--------------------------------------------------------------------------

	/**
	 * @property socket
	 * @since 0.7.0
	 */
	private(set) open var socket: WebSocketConnection!

	//--------------------------------------------------------------------------
	// MARK: Methods
	//--------------------------------------------------------------------------

	/**
	 * @method dispose
	 * @since 0.7.0
	 */
	override open func dispose() {
		self.socket.disconnect()
		self.socket.delegate = nil
		super.dispose()
	}

	//--------------------------------------------------------------------------
	// MARK: JS Functions
	//--------------------------------------------------------------------------

	/**
	 * @method jsFunction_open
	 * @since 0.7.0
	 * @hidden
	 */
	@objc func jsFunction_open(callback: JavaScriptFunctionCallback) {

		if (callback.arguments < 2) {
			fatalError("Method JavaScriptWebSocket.open() requires 2 arguments.")
		}

		let websocketUrl       = callback.argument(0)
		let websocketProtocols = callback.argument(1)

		guard let url = websocketUrl.toURL() else {
			NSLog("Invalid WebSocket URL")
			return
		}

		self.socket = WebSocketConnection(url: url, protocols: websocketProtocols.toArrayOfString())
		self.socket.delegate = self
		self.socket.connect()
	}

	/**
	 * @method jsFunction_send
	 * @since 0.7.0
	 * @hidden
	 */
	@objc func jsFunction_send(callback: JavaScriptFunctionCallback) {

		if (callback.arguments < 1) {
			fatalError("Method JavaScriptWebSocket.send() requires 1 argument.")
		}

		self.socket.write(callback.argument(0).string)
	}

	/**
	 * @method jsFunction_close
	 * @since 0.7.0
	 * @hidden
	 */
	@objc func jsFunction_close(callback: JavaScriptFunctionCallback) {
		self.socket.disconnect()
	}

	//--------------------------------------------------------------------------
	// MARK: WebSocket Delegate
	//--------------------------------------------------------------------------

	/**
	 * @method websocketDidConnect
	 * @since 0.7.0
	 * @hidden
	 */
	open func websocketDidConnect(_ socket: WebSocketConnection) {
		self.callMethod("nativeOnConnect", arguments: [
			self.context.createString(""),
			self.context.createString("")
		])
	}

	/**
	 * @method websocket:didReceiveData
	 * @since 0.7.0
	 * @hidden
	 */
	public func websocket(_ socket: WebSocketConnection, didReceive data: Data?) {
		self.callMethod("nativeOnReceiveData")
	}

	/**
	 * @method websocket:didReceiveMessage
	 * @since 0.7.0
	 * @hidden
	 */
	public func websocket(_ socket: WebSocketConnection, didReceiveMessage string: String) {
		self.callMethod("nativeOnReceiveMessage", arguments: [
			self.context.createString(string)
		])
	}

	/**
	 * @method websocketDidDisconnect
	 * @since 0.7.0
	 * @hidden
	 */
	open func websocketDidDisconnect(_ socket: WebSocketConnection, error: Error?) {
		self.callMethod("nativeOnDisconnect")
	}
}

File: platforms/android/Dezel/dezel/src/main/java/ca/logaritm/dezel/modules/global/JavaScriptWebSocket.kt
package ca.logaritm.dezel.modules.global

import android.os.AsyncTask
import android.os.Handler
import ca.logaritm.dezel.core.JavaScriptClass
import ca.logaritm.dezel.core.JavaScriptContext
import ca.logaritm.dezel.core.JavaScriptFunctionCallback
import ca.logaritm.dezel.extension.fatalError
import com.neovisionaries.ws.client.*

/**
 * @class JavaScriptWebSocket
 * @super JavaScriptClass
 * @since 0.7.0
 */
open class JavaScriptWebSocket(context: JavaScriptContext) : JavaScriptClass(context) {

	//--------------------------------------------------------------------------
	// MARK:Properties
	//--------------------------------------------------------------------------

	/**
	 * @property socket
	 * @since 0.7.0
	 */
	open var socket: WebSocket? = null

	//--------------------------------------------------------------------------
	// Methods
	//--------------------------------------------------------------------------

	/**
	 * @method dispose
	 * @since 0.7.0
	 */
	override fun dispose() {
		this.socket?.disconnect()
		this.socket = null
		super.dispose()
	}

	//--------------------------------------------------------------------------
	// JS Functions
	//--------------------------------------------------------------------------

	/**
	 * @method jsFunction_open
	 * @since 0.7.0
	 * @hidden
	 */
	@Suppress("unused")
	open fun jsFunction_open(callback: JavaScriptFunctionCallback) {

		if (callback.arguments < 2) {
			fatalError("Method JavaScriptWebSocket.open() requires 2 arguments.")
		}

		WebSocketCreator().execute(callback.argument(0).string)
	}

	/**
	 * @method jsFunction_send
	 * @since 0.7.0
	 * @hidden
	 */
	@Suppress("unused")
	open fun jsFunction_send(callback: JavaScriptFunctionCallback) {

		if (callback.arguments < 1) {
			fatalError("Method JavaScriptWebSocket.send() requires 1 argument.")
		}

		this.socket?.sendText(callback.argument(0).string)
	}

	/**
	 * @method jsFunction_close
	 * @since 0.7.0
	 * @hidden
	 */
	@Suppress("unused")
	open fun jsFunction_close(callback: JavaScriptFunctionCallback) {
		this.socket?.disconnect()
	}

	//--------------------------------------------------------------------------
	// WebSocket Handler
	//--------------------------------------------------------------------------

	/**
	 * @method onOpen
	 * @since 0.7.0
	 * @hidden
	 */
	private fun onOpen() {
		this.callMethod("nativeOnConnect", arrayOf(
			this.context.createString(""),
			this.context.createString("")
		))
	}

	/**
	 * @method onClose
	 * @since 0.7.0
	 * @hidden
	 */
	private fun onClose(clean: Boolean) {
		this.callMethod("nativeOnDisconnect")
	}

	/**
	 * @method onMessage
	 * @since 0.7.0
	 * @hidden
	 */
	private fun onMessage(text: String) {
		this.callMethod("nativeOnReceiveMessage", arrayOf(
			this.context.createString(text)
		))
	}

	//--------------------------------------------------------------------------
	// Classes
	//--------------------------------------------------------------------------

	private val handler: Handler = Handler()

	/**
	 * @class WebSocketCreator
	 * @since 0.7.0
	 * @hidden
	 */
	private inner class WebSocketCreator : AsyncTask<String, Void, WebSocket>() {

		/**
		 * @method doInBackground
		 * @since 0.7.0
		 * @hidden
		 */
		override fun doInBackground(vararg params: String?): WebSocket {

			val socket = WebSocketFactory().createSocket(params[0])

			socket.addListener(object : WebSocketAdapter() {

				/**
				 * @method onConnected
				 * @since 0.7.0
				 * @hidden
				 */
				override fun onConnected(websocket: WebSocket?, headers: MutableMap<String, MutableList<String>>?) {
					handler.post {
						onOpen()
					}
				}

				/**
				 * @method onError
				 * @since 0.7.0
				 * @hidden
				 */
				override fun onError(websocket: WebSocket?, cause: WebSocketException?) {
					handler.post {
						onClose(false)
					}
				}

				/**
				 * @method onDisconnected
				 * @since 0.7.0
				 * @hidden
				 */
				override fun onDisconnected(websocket: WebSocket?, serverCloseFrame: WebSocketFrame?, clientCloseFrame: WebSocketFrame?, closedByServer: Boolean) {
					handler.post {
						onClose(closedByServer)
					}
				}

				/**
				 * @method onTextMessage
				 * @since 0.7.0
				 * @hidden
				 */
				override fun onTextMessage(websocket: WebSocket, text: String) {
					handler.post {
						onMessage(text)
					}
				}
			})

			socket.connect()

			return socket
		}

		/**
		 * @method onPostExecute
		 * @since 0.7.0
		 * @hidden
		 */
		override fun onPostExecute(result: WebSocket) {
			socket = result
		}
	}
}